atomic_sequence_test.go

```diff
@@ -0,0 +1,618 @@
+
+	idx, err := seq.Next()
+	if err != nil {
+		t.Error(err.Error())
+	}
+
+	if idx != MaximumBound {
+		t.Error("Could not generate maximum bound value")
+	}
+
+	jdx, err := seq.Next()
+	if err == nil {
+		t.Error("Did not raise exception after going over maximum bound?!")
+	}
+
+	if jdx != 0 {
+		t.Error("returned an overflow value")
+	}
+}
+
+// Test an increment value of 3 with intermediate range.
+func TestIncrementAtomic(t *testing.T) {
+	seq := new(AtomicSequence)
+	seq.Init(3, 10000, 3)
+
+	for i := uint64(3); i < 30; i += 3 {
+		j, _ := seq.Next()
+		if j != i {
+			t.Error("Mismatch counter value during +3 sequence")
+		}
+	}
+}
+
+// Test that sequence goes to the maximum value then errors on increment
+func TestCeilingIncrementAtomic(t *testing.T) {
+	// Create a sequence right at the maximum bound.
+	seq := &AtomicSequence{MaximumBound - 1, 2, MinimumBound, MaximumBound, true}
+
+	jdx, err := seq.Next()
+	if err == nil {
+		t.Error("Did not raise exception after going over maximum bound?!")
+	}
+
+	if jdx != 0 {
+		t.Error("returned an overflow value")
+	}
+}
+
+// Test a maximum value.
+func TestMaximumAtomic(t *testing.T) {
+	seq, err := NewAtomic(32132)
+	if err != nil {
+		t.Error(err.Error())
+	}
+
+	for i := uint64(1); i < 32133; i++ {
+		j, err := seq.Next()
+		if err != nil {
+			t.Error(err.Error())
+		}
+
+		if i != j {
+			t.Error("mismatch during maximum value testing")
+		}
+	}
+
+	for i := 0; i < 10; i++ {
+		val, err := seq.Next()
+		if err == nil {
+			t.Error("should have raised error after maximum reached")
+		}
+		if val != 0 {
+			t.Error("returning non-zero valued response!")
+		}
+	}
+}
+
+// Test a range, note that the monotonically increasing counter range will
+// timeout before tests can be completed.
+func TestRangeAtomic(t *testing.T) {
+	seq, err := NewAtomic(23, 231342)
+	if err != nil {
+		t.Error(err.Error())
+	}
+
+	for i := uint64(23); i < 231343; i++ {
+		j, err := seq.Next()
+		if err != nil {
+			t.Error(err.Error())
+		}
+
+		if i != j {
+			t.Error("mismatch during maximum value testing")
+		}
+	}
+
+	for i := 0; i < 10; i++ {
+		val, err := seq.Next()
+		if err == nil {
+			t.Error("should have raised error after maximum reached")
+		}
+		if val != 0 {
+			t.Error("returning non-zero valued response!")
+		}
+	}
+}
+
+func TestIfAtomicIsSafeForConcurrentUse(t *testing.T) {
+	seq, err := NewAtomic()
+	if err != nil {
+		t.Error(err)
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	go func() {
+		defer wg.Done()
+		var i uint64
+		for ; i < math.MaxUint16; i++ {
+			seq.Next()
+		}
+	}()
+	go func() {
+		defer wg.Done()
+		// keep reading current
+		var i uint64
+		for ; i < math.MaxUint16; i++ {
+			seq.Current()
+		}
+	}()
+
+	wg.Wait()
+}
+
+//===========================================================================
+// Benchmarks
+//===========================================================================
+
+func BenchmarkSequenceAtomic(b *testing.B) {
```

Comment:
I can run this locally, but out of curiosity, how does atomic compare to the thread-unsafe version?

Reply:
`go

goos: linux
goarch: amd64
pkg: github.com/bbengfort/sequence
BenchmarkSequenceAtomic-4   	100000000	        13.0 ns/op	       0 B/op	       0 allocs/op
BenchmarkSequence-4         	1000000000	         2.47 ns/op	       0 B/op	       0 allocs/op
`

It was faster at home (2.08ns/op & 11ns/op in atomic).

Reply:
Awesome thanks - I'm going to add the benchmarks to the README: See #2